### PR TITLE
new_models and updates to the transformer based model build, sweep wi…

### PR DESCRIPTION
**Pull request by oskarbohnlassen**
1. Added an additional dataset
2. Changes to the data.py file to add edge identification with a identity matrix
3. Restructured the model.py file into embeddings, gnn_layers, decoder files instead to setup a more clean way of adding new gnn_layers, encoders etc.
4. Small updates to the train.py file to account for restructuring in (3)
5. Added tests
6. Wandb sweep added

### CODE REVIEW

File: configs/gnn_config_sweep.yaml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    - 0.2
+  batch_size:
+    - 32
+    - 64
+  epochs:
+    - 10
+    - 50
+    - 100
+  weight_decay: 0.0001
+  loss_fn: MSELoss ## L1Loss
```
Comment (fmfsa) [inline configs/gnn_config_sweep.yaml:38]: Why this comment?

File: configs/gnn_config_sweep.yaml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  optimizer: AdamW ## AdamW
```
Comment (fmfsa) [inline configs/gnn_config_sweep.yaml:39]: Same comment as above

File: src/ml_metamodels/gnn_decoders.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+
+class GNNConvDecoder(nn.Module):
+    """Decodes node embeddings and edge features for predictions."""
+
+    def __init__(self, hidden_dim: int) -> None:
+        super().__init__()
+        self.edge_predictor = nn.Linear(hidden_dim * 3, 1)
+
+    def forward(self, x_embeddings, edge_embeddings, edge_index):
```
Comment (fmfsa) [inline src/ml_metamodels/gnn_decoders.py]: So you specified types on __init__ method but not on this one. There's a point to be made about consistency :P I'll let you choose either option

File: configs/gnn_config_sweep.yaml (hunk 1 above)
Comment (oskarbohnlassen) [reply configs/gnn_config_sweep.yaml:38]: Just that this is the other alternative (but let's make more clear in next iteration!

File: src/ml_metamodels/gnn_layers.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
+import torch
+from torch import nn
+from torch_geometric.nn import GCNConv, GATv2Conv
+
+
+class GATConvLayer(nn.Module):
+    """GNN layer for learning node embeddings."""
+
+    def __init__(self, hidden_dim: int, num_gnn_layers: int) -> None:
```
Comment (fmfsa) [inline src/ml_metamodels/gnn_layers.py:9]: same comment about types. Either do it, or don't. But having bad consistency is a bad practice in general. It's destroys readability

File: src/ml_metamodels/gnn_layers.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        self.residual = nn.Linear(hidden_dim, hidden_dim)  # Residual connection
+
+    def forward(self, x, edge_index):
+        for conv in self.convs:
+            x_res = x
+            x = torch.relu(conv(x, edge_index) + self.residual(x_res))  # Update node embeddings
+        return x
+
+
+class GCNConvLayer(nn.Module):
```
Comment (fmfsa) [inline src/ml_metamodels/gnn_layers.py:21]: So this class has exactly the same docstring as the one above. Which is a bad practice. Docstring serves to have a quick way of understanding what it's for. And they say exactly the same thing.

File: src/ml_metamodels/model.py [hunk 6]
```diff
@@ -87,6 +55,44 @@ def forward(self, data: torch.Tensor) -> torch.Tensor:
         return edge_output
 
 
+class DiffusionTestModel(nn.Module):
+    """Simple model for testing purposes."""
+
+    def __init__(self, num_nodes, num_edges):
+        super().__init__()
+
+        # 1) Graph Convolution Layer (as before)
```
Comment (fmfsa) [inline src/ml_metamodels/model.py]: What is "as before"?

File: src/ml_metamodels/train.py [hunk 7]
```diff
@@ -103,11 +103,57 @@ def load_model(self) -> None:
...
+                hidden_dim=self.cfg.model.hidden_dim,
+                num_gnn_layers=self.cfg.model.num_gnn_layers,
+            ).to(self.cfg.train.device)
+
+        elif self.cfg.model.layer_type == "DiffusionTestModel":
+            model = DiffusionTestModel(num_nodes=self.num_nodes, num_edges=self.num_edges).to(self.cfg.train.device)
         else:
             raise ValueError(f"Model type {self.model.layer_type} not supported. Please choose from ['GCN']")
 
+        # Log model to wandb
```
Comment (fmfsa) [inline src/ml_metamodels/train.py:119]: This is the problem with comments and why sometimes it's best not to have them. Where is the code for this comment?

File: src/ml_metamodels/train.py [hunk 8]
```diff
@@ -103,11 +103,57 @@ def load_model(self) -> None:
...
+        print("Model architecture: " + str(model))
+
         return model
 
+    def _get_loss_fn(self) -> torch.nn.Module:
```
Comment (fmfsa) [inline src/ml_metamodels/train.py:124]: About this, I think from a practical point of view, there's no issue I think, but why using L1 vs MSE instead of L1 vs L2 or MAE vs MSE? Because even if they are very similar, they're not exactly the same

File: src/ml_metamodels/train_sweep.py [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+def main() -> None:
+    with initialize(config_path="../../configs"):
+        # hydra.main() decorator was not used since it was conflicting with typer decorator
+        cfg = compose(config_name="gnn_config_sweep.yaml")
+    # Train the model
+    trainer = TrainModelSweep(cfg)
+    trainer.train()
+
+
+@app.command()
```
Comment (fmfsa) [inline src/ml_metamodels/train_sweep.py:335]: Ok cool that you added this to sweep. But then you also need to add the respective part on [project.scripts] and on tasks.py cause otherwise the interaction between typer and invoke is not complete. Let me know if you need more details here

File: src/ml_metamodels/node_embeddings.py [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+    """Encodes node features and edge features."""
+
+    def __init__(self, node_feature_dim: int, edge_feature_dim: int, hidden_dim: int) -> None:
+        super().__init__()
+
+        self.node_encoder = nn.Linear(node_feature_dim, hidden_dim)
+        self.edge_encoder = nn.Linear(edge_feature_dim, hidden_dim)
+
+    def forward(self, x, edge_attr):
+        # Encoded nodes
```
Comment (fmfsa) [inline src/ml_metamodels/node_embeddings.py:15]: This comment adds nothing

File: src/ml_metamodels/node_embeddings.py [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+        encoded_nodes = torch.relu(self.node_encoder(x))
+
+        # Encoded edges
```
Comment (fmfsa) [inline src/ml_metamodels/node_embeddings.py:18]: Same here

File: src/ml_metamodels/node_embeddings.py [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+        encoded_edges = torch.relu(self.edge_encoder(edge_attr))  # Directly use edge_attr
```
Comment (fmfsa) [inline src/ml_metamodels/node_embeddings.py:19]: same about this comment.

File: src/ml_metamodels/train_sweep.py [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+
+app = typer.Typer()
+
+
+class TrainModelSweep:
+    def __init__(self, cfg) -> None:
+        self.cfg = cfg
+        processed_dir = "data/processed"
+        self.processed_data_path = osp.join(processed_dir, cfg.data.dataset_name)
+        self._set_seed(cfg.train.random_seed)  # Set seed for reproducibility
```
Comment (fmfsa) [inline src/ml_metamodels/train_sweep.py:24]: Same with this comment

File: src/ml_metamodels/train_sweep.py [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+
+    def _config_wandb(self) -> None:
+        """Create a logger object."""
+
+        # Load .env file from root directory
```
Comment (fmfsa) [inline src/ml_metamodels/train_sweep.py:29]: Same with this comment

File: src/ml_metamodels/train_sweep.py [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+        load_dotenv(".env")
+        # Example: Access WANDB_API_KEY from the environment
```
Comment (fmfsa) [inline src/ml_metamodels/train_sweep.py:31]: same with this comment :P

File: src/ml_metamodels/train_sweep.py [hunk 16]
```diff
@@ -0,0 +1,366 @@
...
+            model = DiffusionTestModel(num_nodes=self.num_nodes, num_edges=self.num_edges).to(self.cfg.train.device)
+        else:
+            raise ValueError(f"Model type {self.model.layer_type} not supported. Please choose from ['GCN']")
+
+        # Log model to wandb
+        print("Model architecture: " + str(model))
+
+        return model
+
+    def _get_loss_fn(self) -> torch.nn.Module:
```
Comment (fmfsa) [inline src/ml_metamodels/train_sweep.py:145]: We should be careful with code duplication. You already have this code on train.py

File: src/ml_metamodels/train_sweep.py [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+
+        # Ensure split used for training can only be train and validation
+        assert split in ["train", "val"], "Only the splits 'train' and 'val' are supported."
+
+        # Load the data
+        data_path = osp.join(self.processed_data_path, f"{split}.pt")
+        data = torch.load(data_path, weights_only=False)
+        return data
+
+    def load_model(self) -> None:
```
Comment (fmfsa) [inline src/ml_metamodels/train_sweep.py:112]: Same code duplication

**Review by fmfsa — COMMENTED** [review]
Back to you @oskarbohnlassen

File: configs/gnn_config_sweep.yaml (hunk 2 above)
Comment (oskarbohnlassen) [reply configs/gnn_config_sweep.yaml:39]: Here this is the only one I implemented so far!

File: src/ml_metamodels/gnn_decoders.py (hunk 3 above)
Comment (oskarbohnlassen) [reply src/ml_metamodels/gnn_decoders.py]: The curse and blessing of copilot haha but yea we can just remove for now if you prefer to have the consistency in this iteration.

File: src/ml_metamodels/gnn_layers.py (hunk 5 above)
Comment (oskarbohnlassen) [reply src/ml_metamodels/gnn_layers.py:21]: Yea copilot but I mean will just update a bit if you care about it for this PR

File: src/ml_metamodels/train.py (hunk 7 above)
Comment (oskarbohnlassen) [reply src/ml_metamodels/train.py:119]: I think it is logged by the print statement in one of the wandb logs.

File: src/ml_metamodels/train.py (hunk 8 above)
Comment (oskarbohnlassen) [reply src/ml_metamodels/train.py:124]: In the end all of the different loss functions should be implemented. Just picked to and tried to make it work.

File: src/ml_metamodels/train_sweep.py (hunk 17 above)
Comment (oskarbohnlassen) [reply src/ml_metamodels/train_sweep.py:112]: Yea I agree - we should transition into having train_sweep as the only train.py function. However, I wanted to keep both to ensure your cloud integrations were not too impacted. 

But the train_sweep will work just as the train if only one value is given for all the hyperparameters!

File: src/ml_metamodels/train_sweep.py (hunk 9 above)
Comment (oskarbohnlassen) [reply src/ml_metamodels/train_sweep.py:335]: Arh, yeah I did that but didn't copy paste it into this new_new_new_model stuff. Will do!

File: configs/gnn_config_sweep.yaml (hunk 2 above)
Comment (fmfsa) [reply configs/gnn_config_sweep.yaml:39]: I meant, the comment is unecessary and it just bloats the code

File: src/ml_metamodels/gnn_layers.py (hunk 5 above)
Comment (fmfsa) [reply src/ml_metamodels/gnn_layers.py:21]: We can ignore it now and to a profiling thing in a next iteration. But if you decide to do that then maybe add something on MLOPS CANVAS to keep track

File: src/ml_metamodels/train.py (hunk 7 above)
Comment (fmfsa) [reply src/ml_metamodels/train.py:119]: Sure, but that's not what the line below is doing. The line below is printing something. It just so happens somewhere we have wandb configured so that our prints are wandb logs. This is the perfect example why comments are dangerous. Because if now someone comes they will be confused by this comment

File: src/ml_metamodels/train.py (hunk 8 above)
Comment (fmfsa) [reply src/ml_metamodels/train.py:124]: Ah ok got it

File: src/ml_metamodels/train_sweep.py (hunk 9 above)
Comment (fmfsa) [reply src/ml_metamodels/train_sweep.py:335]: Ahahah the infuriating git collapse

File: src/ml_metamodels/train_sweep.py (hunk 17 above)
Comment (fmfsa) [reply src/ml_metamodels/train_sweep.py:112]: So usually if you do this, you should remove the other and verify that the cloud integrations are working. But ok now I get it. We can ignore this, and I will create a PR that cleans this up and migrates all integrations of train.py to train_sweep

Only problem: We should still call it train.py and not train_sweep.py
Would that work for you without destroying any integration with wandb?

Comment (oskarbohnlassen) [reply src/ml_metamodels/train_sweep.py:112]: It is all good calling it train.py! And yes, I understand that it would make sense that I would run it and check that it works with the cloud. 

And the config file name we can also change to gnn_config instead of _sweep.

**Comment by fmfsa** [comment]
👍

**Review by fmfsa — APPROVED** [review]
